File: sangwon/Programmers/92334.cpp
// 프로그래머스 [신고 결과 받기]

#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
#include <unordered_set>

using namespace std;

unordered_map<string, int> countMap;
unordered_map<string, unordered_set<string>> map;

void setHashMap(vector<string> id_list, vector<string> report) {
    for(int i = 0; i <id_list.size(); i++) {
        countMap[id_list[i]] = 0; // 신고횟수 0으로 hashMap 생성
    }
    
    for(int i = 0; i < report.size(); i++) {
        string fUser, sUser;
        int index = report[i].find(' ');
        fUser = report[i].substr(0, index);
        sUser = report[i].substr(index);
        if(map[fUser].find(sUser) == map[fUser].end()) {
            countMap[sUser] += 1; // 신고횟수 증가
            map[fUser].insert(sUser);
        }
    }
}

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    setHashMap(id_list, report);
    vector<int> answer;
    for(int i = 0; i < id_list.size(); i++) {
        int cnt = 0;
        for(auto e : map[id_list[i]]) {
            if(countMap[e] >= k)cnt++;
        }
        answer.push_back(cnt);
    }
    
    return answer;
}

File: sangwon/Programmers/92334.swift
import Foundation

var countMap: [String: Int] = [:] // 닉네임 별 신고횟수를 저장할 딕셔너리
var map: [String: Set<String>] = [:] // 닉네임 별 신고한 id를 저장할 딕셔너리

func makeHashMap(_ id_list:[String], _ report:[String]) {
    for id in id_list {
        countMap[id] = 0
        map[id] = Set()
    }
    for str in report {
        let endIndex = str.index(before: str.endIndex)
        let splitIndex = str.index(after: str.firstIndex(of: " ")!)
        let fUser = String(str[str.startIndex..<str.firstIndex(of: " ")!])
        let sUser = String(str[splitIndex..<str.endIndex])
        // print(fUser, sUser)
        if !map[fUser]!.contains(sUser) {
            map[fUser]?.insert(sUser)
            countMap[sUser]! += 1
        }
        
    }
    // for (key, value) in map {
    //     print("key:", key, "value:", value)
    // }
}

func solution(_ id_list:[String], _ report:[String], _ k:Int) -> [Int] {
    makeHashMap(id_list, report)
    var answer: [Int] = []
    
    // for (key, value) in countMap {
    //     print("key:", key, "value", value)
    // }
    
    for id in id_list {
        var cnt = 0
        for str in map[id]! {
            if countMap[str]! >= k {
                cnt += 1
            }
        }
        answer.append(cnt)
    }
    return answer
}